core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java

```diff
@@ -553,6 +568,128 @@ public SortedRangeSet intersect(ValueSet other)
+                if (startRangeIntersection.isPresent() && endRangeIntersection.isPresent() &&
+                        startRangeIntersection.get().compareTo(startRange) == 0 && endRangeIntersection.get().compareTo(endRange) == 0) {
+                    return probeRangeSet;
+                }
+            }
+            if (blockBuilder == null) {
+                blockBuilder = type.createBlockBuilder(null, 2 * (testEnd + probeEnd));
+                inclusive = new boolean[2 * (testEnd + probeEnd)];
+            }
+            int probeIndex = insertionStartIndex;
+            while (probeIndex < intersectionEndIndex) {
+                RangeView probeRange = probeRangeSet.getRangeView(probeIndex);
+                // intersection at edges as [1, 9], [12, 18] intersected with [7, 15], [17, 21] should end up as [7, 9], [12, 15], [17, 18]
+                if (probeIndex == insertionStartIndex || probeIndex + 1 >= intersectionEndIndex) {
+                    Optional<RangeView> intersect = probeRange.tryIntersect(current);
+                    if (intersect.isPresent()) {
+                        writeRange(type, blockBuilder, inclusive, resultIndex, intersect.get());
+                        resultIndex++;
+                    }
+                    probeIndex++;
+                }
+                else {
+                    Block block = probeRangeSet.getSortedRanges();
+                    if (block instanceof DictionaryBlock || block instanceof ValueBlock) {
+                        int size = intersectionEndIndex - probeIndex - 1;
+                        int offset = probeIndex * 2;
+                        if (block instanceof DictionaryBlock) {
+                            copyDictionaryBlock(blockBuilder, inclusive, probeRangeSet, offset, resultIndex, size);
+                        }
+                        else {
+                            copyValueBlock(blockBuilder, inclusive, probeRangeSet, offset, resultIndex, size);
+                        }
+                        probeIndex += size;
+                        resultIndex += size;
+                    }
+                    else {
+                        // RLE
+                        writeRange(type, blockBuilder, inclusive, resultIndex, probeRange);
+                        resultIndex++;
+                        probeIndex++;
+                    }
+                }
+            }
+        }
+
+        if (blockBuilder == null) {
+            blockBuilder = type.createBlockBuilder(null, 0);
+            inclusive = new boolean[0];
+        }
+        if (resultIndex * 2 < inclusive.length) {
+            inclusive = Arrays.copyOf(inclusive, resultIndex * 2);
+        }
+
+        return new SortedRangeSet(type, inclusive, blockBuilder.build());
+    }
+
+    private static void copyValueBlock(BlockBuilder blockBuilder, boolean[] inclusive, SortedRangeSet source, int sourceOffset, int destinationOffset, int size)
```

Comment:
These parameters don't seem to have any organization (they feel like a direct IntelliJ extraction).  I would order similar to `System.arrayCopy`:
```java
SortedRangeSet source, int sourceOffset, BlockBuilder destination, boolean[] destinationInclusive, int destinationOffset, int size)

Reply:
Also consider unwinding `Source` to the `ValueBlock` and `boolean[] sourceInclusive`